apps/api/src/app/events/usecases/trigger-event/message-matcher.service.ts

```diff
@@ -0,0 +1,266 @@
+    variables: IFilterVariables,
+    configuration?: IMessageFilterConfiguration
+  ): Promise<boolean> {
+    if (!step?.filters || !Array.isArray(step?.filters)) {
+      return true;
+    }
+    if (step.filters?.length) {
+      const foundFilter = await findAsync(step.filters, async (filter) => {
+        const children = filter.children;
+        const noRules = !children || (Array.isArray(children) && children.length === 0);
+        if (noRules) {
+          return true;
+        }
+
+        const singleRule = !children || (Array.isArray(children) && children.length === 1);
+        if (singleRule) {
+          return await this.processFilter(variables, children[0], configuration);
+        }
+
+        return await this.handleGroupFilters(filter, variables, configuration);
+      });
+
+      return foundFilter !== undefined;
+    }
+
+    return true;
+  }
+
+  private async handleGroupFilters(filter, variables: IFilterVariables, configuration: IMessageFilterConfiguration) {
+    if (filter.value === 'OR') {
+      return await this.handleOrFilters(filter, variables, configuration);
+    }
+
+    if (filter.value === 'AND') {
+      return await this.handleAndFilters(filter, variables, configuration);
+    }
+
+    return false;
+  }
+
+  private splitToSyncAsync(filter) {
+    const asyncOnFilters = ['webhook'];
+
+    const asyncFilters = filter.children.filter((childFilter) =>
+      asyncOnFilters.some((asyncOnFilter) => asyncOnFilter === childFilter.on)
+    );
+
+    const syncFilters = filter.children.filter((childFilter) =>
+      asyncOnFilters.some((asyncOnFilter) => asyncOnFilter !== childFilter.on)
+    );
+
+    return { asyncFilters, syncFilters };
+  }
+  private async handleAndFilters(filter, variables: IFilterVariables, configuration: IMessageFilterConfiguration) {
+    const { asyncFilters, syncFilters } = this.splitToSyncAsync(filter);
+
+    const foundSyncFilterMatches = syncFilters.filter((i) => this.processFilterEquality(variables, i));
+    if (syncFilters.length !== foundSyncFilterMatches.length) {
+      return false;
+    }
+
+    const foundAsyncFilterMatches = await filterAsync(asyncFilters, (i) =>
+      this.processFilter(variables, i, configuration)
+    );
+
+    return foundAsyncFilterMatches.length === asyncFilters.length;
+  }
+
+  private async handleOrFilters(filter, variables: IFilterVariables, configuration: IMessageFilterConfiguration) {
+    const { asyncFilters, syncFilters } = this.splitToSyncAsync(filter);
+
+    const syncRes = syncFilters.find((i) => this.processFilterEquality(variables, i));
+    if (syncRes) {
+      return true;
+    }
+
+    return await findAsync(asyncFilters, (i) => this.processFilter(variables, i, configuration));
+  }
+
+  private processFilterEquality(variables: IFilterVariables, i) {
+    const payloadVariable = _.get(variables, [i.on, i.field]);
+    const value = this.parseValue(payloadVariable, i.value);
+    if (i.operator === 'EQUAL') {
```

Comment:
An enum of the operators would be helpful also to provide them in a shared package in order for the users to generate the proper payloads. So it is available in the backend and frontend with the same keys.

Reply:
Same for the logical operators.